### ISSUE TITLE
l1 regularization does nothing for bfgs

### ISSUE THREAD

**Issue by lengstrom** [VowpalWabbit/vowpal_wabbit#356]
Tracing the iterations of the bfgs optimizer with both 0 and 1000000 l1 values does not change the loss function values (at all).

Without any l1 regularization:

```
vw --bfgs --holdout_off --l1 0 --l2 0 -k -d data.tsv -f ./0.hashed_regressor --cache_file ./0.cache --passes 1000 --initial_weight 0 --initial_t 0.01
final_regressor = ./0.hashed_regressor
enabling BFGS based optimization **without** curvature calculation
Num weight bits = 18
learning rate = 0.5
initial_t = 0.01
power_t = 0.5
decay_learning_rate = 1
m = 15
Allocated 34M for weights and mem
## avg. loss    der. mag.   d. m. cond.  wolfe1     wolfe2      mix fraction    curvature   dir. magnitude  step size 
creating cache_file = ./0.cache
Reading datafile = data.tsv
num sources = 1
 1 39.75810     2303.79890  2238.75861                                      4235207.68967   2202.94775  0.00053   
 3 39.16639     4872.26846  4671.52333   0.500001   -0.000001                           0.00042     1.00000   
 4 38.37624     4635.16984  4444.29911   0.987664   0.975326                            0.88780     1.00000   
 5 19.16637     367.16732   359.86364    0.540851   0.081738                            0.04325     1.00000   
 6 16.78974     269.65303   264.38808    0.845747   0.684647                            0.45597     1.00000   
 7 12.24979     164.48068   158.24200    0.548937   0.100427                            0.22044     1.00000   
 8 12.08322     545.38933   527.40834    0.041493   -0.929687                           0.02195     1.00000   
 9 10.44693     64.98041    63.27632     0.601153   0.222400                            0.01244     1.00000   
10 9.84662      31.35905    30.73519     0.814216   0.628441                            0.14609     1.00000   
11 8.66950      62.17668    60.43105     0.699023   0.398056                            0.22581     1.00000   
12 7.46970      71.97469    69.85544     0.764919   0.529820                            1.14852     1.00000   
13 5.72989      129.95964   126.19237    0.522197   0.044467                            0.96975     1.00000   
14 6.82936      642.89861   620.16381    -0.329352  -1.705646                           (revise x 0.5)  0.50000   
15 5.19524      132.28156   127.82354    0.320314   -0.371646                           0.01028     1.00000
```

With l1 regularization:

```
vw --bfgs --holdout_off --l1 1000000 --l2 0 -k -d data.tsv -f ./0.hashed_regressor --cache_file ./0.cache --passes 1000 --initial_weight 0 --initial_t 0.01
using l1 regularization = 1e+06
final_regressor = ./0.hashed_regressor
enabling BFGS based optimization **without** curvature calculation
Num weight bits = 18
learning rate = 0.5
initial_t = 0.01
power_t = 0.5
decay_learning_rate = 1
m = 15
Allocated 34M for weights and mem
## avg. loss    der. mag.   d. m. cond.  wolfe1     wolfe2      mix fraction    curvature   dir. magnitude  step size 
creating cache_file = ./0.cache
Reading datafile = data.tsv
num sources = 1
 1 39.75810     2303.79890  2238.75861                                      4235207.68967   2202.94775  0.00053   
 3 39.16639     4872.26846  4671.52333   0.500001   -0.000001                           0.00042     1.00000   
 4 38.37624     4635.16984  4444.29911   0.987664   0.975326                            0.88780     1.00000   
 5 19.16637     367.16732   359.86364    0.540851   0.081738                            0.04325     1.00000   
 6 16.78974     269.65303   264.38808    0.845747   0.684647                            0.45597     1.00000   
 7 12.24979     164.48068   158.24200    0.548937   0.100427                            0.22044     1.00000   
 8 12.08322     545.38933   527.40834    0.041493   -0.929687                           0.02195     1.00000   
 9 10.44693     64.98041    63.27632     0.601153   0.222400                            0.01244     1.00000   
10 9.84662      31.35905    30.73519     0.814216   0.628441                            0.14609     1.00000   
11 8.66950      62.17668    60.43105     0.699023   0.398056                            0.22581     1.00000   
12 7.46970      71.97469    69.85544     0.764919   0.529820                            1.14852     1.00000   
13 5.72989      129.95964   126.19237    0.522197   0.044467                            0.96975     1.00000   
14 6.82936      642.89861   620.16381    -0.329352  -1.705646                           (revise x 0.5)  0.50000   
15 5.19524      132.28156   127.82354    0.320314   -0.371646                           0.01028     1.00000   
```
